### ISSUE TITLE
Can't save and reload model that contains a text.keras.layers.ToDense layer

### ISSUE THREAD

**Issue by jasonbrancazio** [tensorflow/text#271]
An LSTM model I created inspired by https://github.com/tensorflow/text/blob/master/examples/keras_example_174.ipynb can be successfully trained. However, I get an error when trying to save and reload it.

It turns out that even the toy model in the docstring for text.keras.layers.ToDense can't be saved and reloaded.

Using either 2.1.1 or 2.2.0-rc2:
```
x = tf.keras.layers.Input(shape=(None, None), ragged=True)
y = text.keras.layers.ToDense(mask=True)(x)
test_model = tf.keras.Model(x, y)
test_model.save('/tmp/test_model')
reloaded_test_model = tf.keras.models.load_model('/tmp/test_model')
```

Error:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-990dc0115c64> in <module>()
      3 test_model = tf.keras.Model(x, y)
      4 test_model.save('/tmp/test_model')
----> 5 reloaded_test_model = tf.keras.models.load_model('/tmp/test_model')

11 frames
/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/saving/save.py in load_model(filepath, custom_objects, compile)
    188     if isinstance(filepath, six.string_types):
    189       loader_impl.parse_saved_model(filepath)
--> 190       return saved_model_load.load(filepath, compile)
    191 
    192   raise IOError(

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/saving/saved_model/load.py in load(path, compile)
    114   # TODO(kathywu): Add saving/loading of optimizer, compiled losses and metrics.
    115   # TODO(kathywu): Add code to load from objects that contain all endpoints
--> 116   model = tf_load.load_internal(path, loader_cls=KerasObjectLoader)
    117 
    118   # pylint: disable=protected-access

/usr/local/lib/python3.6/dist-packages/tensorflow/python/saved_model/load.py in load_internal(export_dir, tags, loader_cls)
    602       loader = loader_cls(object_graph_proto,
    603                           saved_model_proto,
--> 604                           export_dir)
    605       root = loader.get(0)
    606       if isinstance(loader, Loader):

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/saving/saved_model/load.py in __init__(self, *args, **kwargs)
    186     self._models_to_reconstruct = []
    187 
--> 188     super(KerasObjectLoader, self).__init__(*args, **kwargs)
    189 
    190     # Now that the node object has been fully loaded, and the checkpoint has

/usr/local/lib/python3.6/dist-packages/tensorflow/python/saved_model/load.py in __init__(self, object_graph_proto, saved_model_proto, export_dir)
    121       self._concrete_functions[name] = _WrapperFunction(concrete_function)
    122 
--> 123     self._load_all()
    124     self._restore_checkpoint()
    125 

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/saving/saved_model/load.py in _load_all(self)
    213 
    214     # Finish setting up layers and models. See function docstring for more info.
--> 215     self._finalize_objects()
    216 
    217   @property

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/saving/saved_model/load.py in _finalize_objects(self)
    508 
    509     # Initialize graph networks, now that layer dependencies have been resolved.
--> 510     self._reconstruct_all_models()
    511 
    512   def _unblock_model_reconstruction(self, layer_id, layer):

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/saving/saved_model/load.py in _reconstruct_all_models(self)
    526       all_initialized_models.add(model_id)
    527       model, layers = self.model_layer_dependencies[model_id]
--> 528       self._reconstruct_model(model_id, model, layers)
    529       self._add_object_graph_edges(self._proto.nodes[model_id], model_id)
    530       _finalize_config_layers([model])

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/saving/saved_model/load.py in _reconstruct_model(self, model_id, model, layers)
    562     else:
    563       (inputs, outputs, created_layers) = network_lib.reconstruct_from_config(
--> 564           config, created_layers={layer.name: layer for layer in layers})
    565       model.__init__(inputs, outputs, name=config['name'])
    566       network_lib.connect_ancillary_layers(model, created_layers)

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/network.py in reconstruct_from_config(config, custom_objects, created_layers)
   2022       if layer in unprocessed_nodes:
   2023         for node_data in unprocessed_nodes.pop(layer):
-> 2024           process_node(layer, node_data)
   2025 
   2026   input_tensors = []

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/network.py in process_node(layer, node_data)
   1970     if input_tensors is not None:
   1971       input_tensors = base_layer_utils.unnest_if_single_tensor(input_tensors)
-> 1972       output_tensors = layer(input_tensors, **kwargs)
   1973 
   1974       # Update node index map.

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/base_layer.py in __call__(self, *args, **kwargs)
    889           raise ValueError('Layer %s does not support RaggedTensors as input. '
    890                            'Inputs received: %s. You can try converting your '
--> 891                            'input to an uniform tensor.' % (self.name, inputs))
    892 
    893         graph = backend.get_graph()

ValueError: Layer to_dense does not support RaggedTensors as input. Inputs received: tf.RaggedTensor(values=tf.RaggedTensor(values=Tensor("Placeholder_3:0", shape=(None,), dtype=float32), row_splits=Tensor("Placeholder_5:0", shape=(None,), dtype=int64)), row_splits=Tensor("Placeholder_4:0", shape=(None,), dtype=int64)). You can try converting your input to an uniform tensor.
```

It's strange that the error upon reloading reports that the ToDense layer does not support RaggedTensors as input. The __init__ for the layer sets _supports_ragged_inputs to True.
https://github.com/tensorflow/text/blob/master/tensorflow_text/python/keras/layers/todense.py#L75

**Comment by jasonbrancazio** [comment]
Update: I believe it is due to the lack of _supports_ragged_inputs in the Masking layer. If you replace the text.keras.layers.ToDense in the previous example with some of the procedures in its source code, you get a more informative error:

```
x = tf.keras.layers.Input(shape=(None, None), ragged=True)
dense_x = x.to_tensor(default_value=0)
y = tf.keras.layers.Masking(mask_value=0)(x)
test_model = tf.keras.Model(x, y)
test_model.save('/tmp/test_model')
reloaded_test_model = tf.keras.models.load_model('/tmp/test_model')
```

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-12-9086c5e1c3fb> in <module>()
      1 x = tf.keras.layers.Input(shape=(None, None), ragged=True)
      2 dense_x = x.to_tensor(default_value=0)
----> 3 y = tf.keras.layers.Masking(mask_value=0)(x)
      4 test_model = tf.keras.Model(x, y)
      5 test_model.save('/tmp/test_model')

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/base_layer.py in __call__(self, *args, **kwargs)
    889           raise ValueError('Layer %s does not support RaggedTensors as input. '
    890                            'Inputs received: %s. You can try converting your '
--> 891                            'input to an uniform tensor.' % (self.name, inputs))
    892 
    893         graph = backend.get_graph()

ValueError: Layer masking_4 does not support RaggedTensors as input. Inputs received: tf.RaggedTensor(values=tf.RaggedTensor(values=Tensor("Placeholder_15:0", shape=(None,), dtype=float32), row_splits=Tensor("Placeholder_17:0", shape=(None,), dtype=int64)), row_splits=Tensor("Placeholder_16:0", shape=(None,), dtype=int64)). You can try converting your input to an uniform tensor.

**Comment by jasonbrancazio** [comment]
@broken would you mind taking a look? Before refactoring my model, I'd love to know if the inability to save and reload a text.keras.layers.ToDense layer is a known issue. Are there any known workarounds?

**Comment by markomernick** [comment]
Hey @jasonbrancazio, thanks for this report! I will route this to the right folks.

**Comment by jasonbrancazio** [comment]
I tried adding `self._supports_ragged_inputs = True` to a custom version of the Masking layer that is used by the ToDense layer, but that didn't work.

**Comment by jasonbrancazio** [comment]
Prior to saving the toy model from the README, I can verify that it still believes the ToDense layer supports ragged inputs by observing this `test_model.layers[1]._supports_ragged_inputs` equals True. Something about saving or reloading the model seems to be ignoring that property.

**Comment by markomernick** [comment]
Hey @jasonbrancazio -

We're going to be removing the supports_ragged_inputs attr. This change is in internally and should show up in the nightlies as soon as it is sync'd.

**Comment by jasonbrancazio** [comment]
@markomernick thanks for the update. Will it make it into the 2.2 release?

**Comment by omalleyt12** [comment]
@jasonbrancazio This fix unfortunately won't make it into the 2.2 release, at this point we're only accepting security fixes or fixes for major regressions, and this issue existed in prior version

**Comment by everdark** [comment]
Just got this issue today.
Is there any workaround before the fix? My intention is to use `RaggedTensor` as model inputs.
